Automattic/wp-calypso
# Masterbar - always show site items (refactor `prevSelectedSiteId` -> `mostRecentlySelectedSiteId`)

https://github.com/Automattic/wp-calypso/pull/92800

## Proposed Changes

- This PR changes `prevSelectedSiteId` to `mostRecentlySelectedSiteId` to be updated simultaneously and ensures that `prevSelectedSiteId` is not to be null. Please see the `Why are these changes being made?` section. 
- Renamed `prevSelectedSiteId` -> `mostRecentlySelectedSiteId` so that it represents what it actually does. 

## Why are these changes being made?


I noticed a logic that is likely to cause tricky behavior in the future when I was testing https://github.com/Automattic/wp-calypso/pull/92800. Let's consider the following steps;

1. Land on /sites
1. Select a site A
1. Select a different site B
1. Observe the site B is selected in the masterbar
	- Here, the redux state _should_ hold: `current=site B` and `previous=site A`, but it actually hold `current=site B` and `previous=site B`
1. Click "Themes"
	- Here, the redux state _should_ hold: `current=null` and `previous=site B` since going to /themes [clears](https://github.com/Automattic/wp-calypso/blob/f93f63615f047abe2a07f1028b049abce2e1fede/client/my-sites/themes/controller.jsx#L67) the selected site.
1. Observe the site B is selected in the masterbar

This result is not a bug at all from a behavioral point of view, but I noticed in step 5 that the currently selected site and the previously selected site were the same. This is not expected from a programmatic point of view, as the name `prevSelectedSiteId` suggests. 
So, what was happening?: It was actually executing `setSelectedSiteId` multiple times, so the currently selected site and the previously selected site became the same eventually;

- Initial: `current=null` and `previous=null`
- Selecting site A fires `setSelectedSiteId` multiple times: 
	- 1: `current=site A` and `previous=null`
	- 2: `current=site A` and `previous=site A`
	- 3: `current=site A` and `previous=site A`
	- ...

It seemed to be executed in https://github.com/Automattic/wp-calypso/blob/f93f63615f047abe2a07f1028b049abce2e1fede/client/hosting/sites/hooks/use-sync-selected-site.ts#L23 and https://github.com/Automattic/wp-calypso/blob/f93f63615f047abe2a07f1028b049abce2e1fede/client/my-sites/controller.js#L608 with one selection of a site.

<img width="1016" alt="Screenshot 2024-07-19 at 11 45 05" src="https://github.com/user-attachments/assets/015f5bd3-97d2-4828-a32e-3fa5f48a32fe">

The logic worked since it executes `setSelectedSiteId` only once in step **6**. If we somehow change to call `setSelectedSiteId` multiple times in the future, the previously selected site will become null.
I considered limiting this multiply-executed `setSelectedSiteId`, but I preferred guarding with a non-null value would be more reliable for future changes. Additionally, I think it does not necessarily represent the "previously" selected site, so I changed it to the most recently selected site. 

## Testing Instructions



* Go to calypso /sites and select a site to overview.
* Verify the site menu, +new,  etc; links continue to work as expected.
* Close the overview tab and verify that the site menu does not disappear.
* Verify the site menu is still present (previously would disappear at this point) and that, +new, etc; links continue to work as expected.
* Click on the domains tab on the global sidebar. Verify that the site-specific items are still on the global sidebar. verify the masterbar and that this domains page works as expected.
* Do the above for global plugins and themes pages.
* Smoke test around calypso. verify things work as expected, and we haven't created weird edge cases.

## Pre-merge Checklist
